lib/client/keystore.go

```diff
@@ -687,3 +693,65 @@ func (noLocalKeyStore) SaveCerts(proxy string, cas []auth.TrustedCerts) error {
 }
 func (noLocalKeyStore) GetCerts(proxy string) (*x509.CertPool, error) { return nil, errNoLocalKeyStore }
 func (noLocalKeyStore) GetCertsPEM(proxy string) ([][]byte, error)    { return nil, errNoLocalKeyStore }
+
+type usernameProxyPair struct {
+	username string
+	proxy    string
+}
+
+// MemLocalKeyStore is an in-memory session keystore implementation.
+type MemLocalKeyStore struct {
+	fsLocalNonSessionKeyStore
+	inMem map[usernameProxyPair]*Key
+}
+
+// NewMemLocalKeyStore initializes a MemLocalKeyStore, the key directory here is only used
+// for storing CA certificates and known host fingerprints.
+func NewMemLocalKeyStore(dirPath string) (*MemLocalKeyStore, error) {
+	dirPath, err := initKeysDir(dirPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	inMem := make(map[usernameProxyPair]*Key)
+	return &MemLocalKeyStore{fsLocalNonSessionKeyStore: fsLocalNonSessionKeyStore{
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: teleport.ComponentKeyStore,
+		}),
+		KeyDir: dirPath,
+	}, inMem: inMem}, nil
+}
+
+// AddKey writes a key to the underlying key store if MemLocalKeyStore was initialized with saveNewKeys set to true.
+func (s *MemLocalKeyStore) AddKey(proxy string, username string, key *Key) error {
+	s.inMem[usernameProxyPair{username, proxy}] = key
+	return nil
+}
+
+// GetKey returns the session key for the given username and proxy.
+func (s *MemLocalKeyStore) GetKey(proxy, username string, opts ...KeyOption) (*Key, error) {
+	entry, ok := s.inMem[usernameProxyPair{username, proxy}]
+	if !ok {
+		return nil, trace.NotFound("key for proxy: %q and username: %q not found", proxy, username)
+	}
+
+	return entry, nil
+}
```

Comment:
Why are `opts` ignored? Note that k8s/DB access assumes to be able to fetch their certs per cluster (which is how the certs are cached within the FS store). That's how SSH certs are going to be stored/fetched after #5938 as well.

Since `inMem` is indexed just by `username, proxy` (rather than `username, proxy, cluster`) it seems that `GetKey` would always return the certs that happened to be added with the last `AddKey` call, thus ignoring the cluster that may have been explicitly requested, is that correct?

Even if nothing were strictly speaking broken by such behaviour, it could cause unnecessary load on the root auth/proxy servers inasmuch as the per-cluster certs would have to be reissued more often.

Reply:
Ah, yeah, you're correct. I'll take care of options and add an explicit cluster name parameter and index by it.

Reply:
Or would it perhaps be better if I just rebase on your PR when it's merged? Is it in a state to be merged relatively soon? i.e today or early next week?